libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -0,0 +1,296 @@
+import com.mapbox.android.telemetry.MapboxTelemetry
+import com.mapbox.navigation.base.exceptions.NavigationException
+import com.mapbox.navigation.base.extensions.ifNonNull
+import com.mapbox.navigation.base.logger.model.Message
+import com.mapbox.navigation.base.logger.model.Tag
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.BuildConfig
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.MapboxNavigationOptions
+import com.mapbox.navigation.core.trip.session.RouteProgressObserver
+import com.mapbox.navigation.logger.MapboxLogger
+import com.mapbox.navigation.utils.thread.ThreadController
+import com.mapbox.navigation.utils.thread.monitorChannelWithException
+import java.util.Locale
+import java.util.UUID
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.selects.select
+import kotlinx.coroutines.withContext
+
+private enum class LocationBufferCommands {
+    BUFFER_ADD,
+    BUFFER_REMOVE,
+    BUFFER_CLEAR,
+    BUFFER_READ,
+}
+
+/**
+ * This data class defines an action to be taken against the location buffer. These can be one of four defined in LocationBufferCommands
+ */
+private data class LocationBufferControl(val command: LocationBufferCommands, val location: Location?, val predicate: ((Location) -> Unit) = { _ -> Unit })
+
+internal object MapboxNavigationTelemetry : MapboxNavigationTelemetryInterface {
+    private lateinit var context: Context
+    private lateinit var mapboxToken: String
+    private const val MAPBOX_NAVIGATION_USER_AGENT_BASE = "mapbox-navigation-android"
+    private const val MAPBOX_NAVIGATION_UI_USER_AGENT_BASE = "mapbox-navigation-ui-android"
+
+    private val TAG = Tag("MAPBOX_TELEMETRY")
+    private const val MAX_LOCATION_VALUES = 20
+    private const val MAX_TELEMTRY_EVENTS = 100
+    private const val ONE_SECOND = 1000L
+    private val telemetryEventsQueue = mutableListOf<TelemetryEventInterface>()
+    private val jobControl = ThreadController.getIOScopeAndRootJob()
+    private val locationBuffer = mutableListOf<Location>()
+    private val channelLocationBuffer = Channel<LocationBufferControl>(MAX_LOCATION_VALUES)
+    private val channelOnRouteProgress = Channel<RouteProgress>(Channel.CONFLATED) // we want just the last notification
+    private lateinit var cleanupJob: Job
+    private lateinit var mapboxTelemetry: MapboxTelemetry
+    private lateinit var locationEngine: LocationEngine
+
+    // Location request settings
+    private val locationEngineRequest: LocationEngineRequest = LocationEngineRequest.Builder(ONE_SECOND)
+            .setPriority(LocationEngineRequest.PRIORITY_HIGH_ACCURACY)
+            .build()
+
+    // Call back that receives
+    private val routeProgressListener = object : RouteProgressObserver {
```

Comment:
Wondering if we should listen for route progress updates from `MapboxNavigation` and send the updates to here `MapboxNavigationTelemetry` 🤔 What do you think? Same applies to location updates.

Reply:
I agree, and preferably it should already be parsed into whatever format the telemetry wrappers expect. This way, we're removing the cross dependency.

Reply:
Still valid.

Reply:
👍 done